### PR TITLE
[Agent] Support for config constraints

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

Followup tohttps://github.com/elastic/beats/issues/15690. 
This PR introduces contraints as described in this issue: https://github.com/elastic/beats/issues/16409

This PR enhances emitter not only to use decorators to enrich configuration but filters to modify ast by dropping some nodes. 

It introduces also constraints filter which loads all datasets and for those which contain constraint node parses this nodes and uses existing boolexp implementation to evaluate expressions. 
It also allows user to use some basic variables such as `agent.id`, `agent.version` and such.

Nodes containing constraints section and where constraint is evaluated as FALSE, these nodes are dropped.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works

Fixes: #16409

### CODE REVIEW

File: x-pack/agent/pkg/agent/application/emitter.go [hunk 1]
```diff
@@ -33,6 +39,12 @@ func emitter(log *logger.Logger, router *router, decorators ...decoratorFunc) em
 			return errors.New(err, "could not create the AST from the configuration", errors.TypeConfig)
 		}
 
+		for _, filter := range modifiers.Filters {
+			if err := filter(ast); err != nil {
+				return errors.New(err, "failed to filter configuration", errors.TypeConfig)
+			}
+		}
```
Comment (ph) [inline x-pack/agent/pkg/agent/application/emitter.go]: Maybe we should log when a constrains fails?

File: x-pack/agent/pkg/agent/application/filters/constraints_filter.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+const (
+	agentIDKey      = "agent.id"
+	agentVersionKey = "agent.version"
+	hostArchKey     = "host.architecture"
+	osFamilyKey     = "os.family"
+	osKernelKey     = "os.kernel"
+	osPlatformKey   = "os.platform"
+	osVersionKey    = "os.version"
+)
```
Comment (ph) [inline x-pack/agent/pkg/agent/application/filters/constraints_filter.go]: Can you add a documentation with the variables that are accessible in the configuration?
This will be useful when writing the documentation and also when we implement them in UI.

File: x-pack/agent/pkg/agent/application/filters/constraints_filter.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	}
+
+	return true, nil
+}
+
+func evaluateConstraint(constraint string) (bool, error) {
+	regs := boolexp.NewMethodsReg()
+	if err := regs.Register(validateVersionFuncName, regValidateVersion); err != nil {
+		return false, err
+	}
```
Comment (ph) [inline x-pack/agent/pkg/agent/application/filters/constraints_filter.go]: This could be defined once and reused?

File: x-pack/agent/pkg/agent/application/filters/constraints_filter.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+	store, err := newVarStore()
+	if err != nil {
+		return false, err
+	}
+
+	if err := initVarStore(store); err != nil {
```
Comment (ph) [inline x-pack/agent/pkg/agent/application/filters/constraints_filter.go]: Looking a the fields used I think this could be defined once and reused?

The following requires a reboot to change:
```
	// 	Agent
	store.vars[agentIDKey] = agentInfo.AgentID()
	store.vars[agentVersionKey] = release.Version()

	// Host
	store.vars[hostArchKey] = info.Architecture

	// Operating system
	store.vars[osFamilyKey] = runtime.GOOS
	store.vars[osKernelKey] = info.KernelVersion
	store.vars[osPlatformKey] = info.OS.Family
	store.vars[osVersionKey] = info.OS.Version

File: x-pack/agent/pkg/boolexp/boolexp_test.go [hunk 5]
```diff
@@ -36,6 +36,7 @@ func TestBoolexp(t *testing.T) {
 	}{
 		// Variables
 		{expression: "%{[hello.var]} == 'hello'", result: true},
+		{expression: "%{[hello.var]} != 'hello'", result: false},
```
Comment (ph) [inline x-pack/agent/pkg/boolexp/boolexp_test.go]: oops I've missed that :)

File: x-pack/agent/pkg/agent/application/emitter.go (hunk 1 above)
Comment (ph) [reply x-pack/agent/pkg/agent/application/emitter.go]: I think I did not correctly express my requirement here, my bad. 

The above code do log when a constraint fails, but I think we should also log when a constraints matches and we filter content.

File: x-pack/agent/pkg/agent/application/filters/constraints_filter.go [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+		strval, ok := c.(*transpiler.StrVal)
+		if !ok {
+			return false, errors.New("constraints is not a string")
+		}
+
+		constraint := strval.String()
+		if isOK, err := evaluateConstraint(constraint); !isOK || err != nil {
+			if err == nil {
+				// log only constraint not matching
+				log.Infof("constraint '%s' not matching for datasource: %v", constraint, datasourceNode.String())
```
Comment (ph) [inline x-pack/agent/pkg/agent/application/filters/constraints_filter.go]: We might want to reduce what we want to output to the log to the type of the **datasource** and **id**? This is just to make sure we do not expose credentials in the log this could be a problem with metricbeat where datasource have a higher change of containing secrets.

File: x-pack/agent/pkg/agent/application/info/agent_id.go [hunk 7]
```diff
@@ -80,6 +80,10 @@ func getInfoFromStore(s ioStore) (*persistentAgentInfo, error) {
 			errors.M(errors.MetaKeyPath, AgentConfigFile))
 	}
 
+	if err := reader.Close(); err != nil {
```
Comment (ruflin) [inline x-pack/agent/pkg/agent/application/info/agent_id.go]: How is this related to this PR?

File: x-pack/agent/pkg/agent/application/info/agent_id.go [hunk 8]
```diff
@@ -118,6 +122,10 @@ func updateAgentInfo(s ioStore, agentInfo *persistentAgentInfo) error {
...
```
Comment (ruflin) [inline x-pack/agent/pkg/agent/application/info/agent_id.go]: How is this related to this PR?

File: x-pack/agent/pkg/agent/application/info/agent_id.go (hunk 7 above)
Comment (michalpristas) [reply x-pack/agent/pkg/agent/application/info/agent_id.go]: this fixes an issue on windows i experienced on tests by introducing more extensive use of agentInfo.ID()

File: x-pack/agent/pkg/agent/application/info/agent_id.go (hunk 8 above)
Comment (michalpristas) [reply x-pack/agent/pkg/agent/application/info/agent_id.go]: same as ^^^

File: x-pack/agent/pkg/agent/application/filters/constraints_filter.go (hunk 6 above)
Comment (michalpristas) [reply x-pack/agent/pkg/agent/application/filters/constraints_filter.go]: constriant is at datasource level and as such it does not contain type, ID is optional for now. i think namespace, output is a identifying configuration and id if present will be logged

Comment (ph) [reply x-pack/agent/pkg/agent/application/filters/constraints_filter.go]: sound good ^

**Comment by ph** [comment]
> Should this have somewhere a changelog entry?

> Could you update the PR description with more details on what is actually done in this PR and how? At the moment it references to 3 different issues.

Good comment, I've missed that.. I need to update my local checklist for PR review.
